spatialdata/_core/_spatialdata.py

```diff
@@ -0,0 +1,262 @@
+    points: Mapping[str, AnnData] = MappingProxyType({})
+    polygons: Mapping[str, GeoDataFrame] = MappingProxyType({})
+    shapes: Mapping[str, AnnData] = MappingProxyType({})
+    _table: Optional[AnnData] = None
+
+    def __init__(
+        self,
+        images: Mapping[str, Any] = MappingProxyType({}),
+        labels: Mapping[str, Any] = MappingProxyType({}),
+        points: Mapping[str, Any] = MappingProxyType({}),
+        polygons: Mapping[str, Any] = MappingProxyType({}),
+        shapes: Mapping[str, Any] = MappingProxyType({}),
+        table: Optional[AnnData] = None,
+        **kwargs: Any,
+    ) -> None:
+
+        if images is not None:
+            self.images = {}
+            for k, v in images.items():
+                if ndim(v) == 3:
+                    Image2D_s.validate(v)
+                    self.images[k] = v
+                elif ndim(v) == 4:
+                    Image3D_s.validate(v)
+                    self.images[k] = v
+
+        if labels is not None:
+            self.labels = {}
+            for k, v in labels.items():
+                if ndim(v) == 2:
+                    Label2d_s.validate(v)
+                    self.labels[k] = v
+                elif ndim(v) == 3:
+                    Label3D_s.validate(v)
+                    self.labels[k] = v
+
+        if polygons is not None:
+            self.polygons = {}
+            for k, v in polygons.items():
+                Polygon_s.validate(v)
+                self.polygons[k] = v
+
+        if shapes is not None:
+            self.shapes = {}
+            for k, v in shapes.items():
+                Shape_s.validate(v)
+                self.shapes[k] = v
+
+        if points is not None:
+            self.points = {}
+            for k, v in points.items():
+                Point_s.validate(v)
+                self.points[k] = v
+
+        if table is not None:
+            Table_s.validate(table)
+            self._table = table
+
+    def write(self, file_path: str) -> None:
+        """Write to Zarr file."""
+
+        store = parse_url(file_path, mode="w").store
+        root = zarr.group(store=store)
+
+        # get union of unique ids of all elements
+        elems = set().union(
+            *[
+                set(i)
+                for i in [
+                    self.images,
+                    self.labels,
+                    self.points,
+                    self.polygons,
+                    self.shapes,
+                ]
+            ]
+        )
+
+        for el in elems:
+            elem_group = root.create_group(name=el)
+            if self.images is not None and el in self.images.keys():
+                write_image(
+                    image=self.images[el],
+                    group=elem_group,
+                    name=el,
+                    storage_options={"compressor": None},
+                )
+            if self.labels is not None and el in self.labels.keys():
```

Comment:
can self.labels be None?

Reply:
(same for all elements)

Reply:
As the code is written now no, it would likely break things, but the constructors for the elements assures that there is always an empty dict when elements are not specified.

Reply:
In that case, in this we should remove the `if * is None` condition.

Reply:
done thanks